modules/README.md

```diff
@@ -115,35 +119,235 @@ Kudos to @mfw78
+
+It is important to understand the signing and verification flow related to ERC-1271. Any dapp can make use of `eth_sign` or `eth_signTypedData` to retrieve a signature from a connected wallet. To verify this signature for an EOA the dapp will calculate the hash of the message (either ERC-191 or ERC-712 based) and recover the signer address from the signature using the ECDSA based recovery. For smart contracts the flow is similar, but instead of using the ECDSA based recovery method to validate the signer, the dapp calls `isValidSignature` on the smart contract with the hash and the signature.
+
+This allows us to add additional logic when validating signatures, such as differentiating the validation logic based on the message. The challenge is that it is impossible to retrieve the message that was signed from the hash. To work around it we can encode the message into the signature and then when the smart contract is called this message is extracted, hashed and validated against the hash. This is outlined in the following example code:
+
+```solidity
+function isValidSignature(bytes32 hash, bytes memory payload) external view returns (bytes4) {
+  (bytes memory signature, bytes memory message) = abi.decode(payload, (bytes,bytes));
+  require(hash == keccak256(message), "Unexpected message/hash combination");
+  // more validation logic
+}
+```
+
+Using this approach it is also possible to encode ERC-712 based messages in the signature (for signatures requested via `eth_signTypedData`). With this we can route the signature validation based on the domain of the ERC-712 message. The payload contain the signature, the domainHash and the hashStruct. Only minor adjustments are required to the previous code:
+
+```solidity
+mapping(bytes32 => ISignatureValidator) validatorForDomain;
+function isValidSignature(bytes32 hash, bytes memory payload) external view returns (bytes4) {
+  (bytes32 domainHash, bytes32 hashStruct, bytes memory signature) = abi.decode(payload, (bytes32,bytes32,bytes));
+  require(hash, keccak256(0x19, 0x01, domainHash, hashStruct), "Could not validate data");
+  ISignatureValidator validator = validatorForDomain[domainHash];
+  return validator.isValidSignature(msg.sender /* account */, _msgSender(), domainHash, hashStruct, signature);
+}
+```
+
+To distinguish signatures that follow this format from "default" signatures the signature is prepended with a special bytes4 identifier. When the dapp requests the signature the wallet can then create the signature in the required format for the correct signature routing.
+
+The encoded data to be received by the signature validator manager as `bytes` parameter of `isValidSignature(bytes32,bytes)` is expected to be encoded as follows:
+
+```solidity
+function encodeData(bytes4 selector, bytes32 domain, bytes32 hashStruct, bytes calldata signatures) public pure returns (bytes memory) {
```

Comment:
what is the selector? 

Also isn't this the same as `abi.encodeWithSelector`? 

Could we use `abi.encodeWithSignature` and define it like a method call?